File: Day19.py
import functools


def get_file_data(file_name):
    f = open(file_name)
    data = []
    for line in f:
        data.append(line.rstrip())
    return data


@functools.cache
def is_possible(towel):
    if towel == '':
        return True, 1
    possible = False
    count = 0
    for start in available_designs:
        if towel.startswith(start):
            check, number = is_possible(towel[len(start):])
            possible = possible or check
            count += number
    return possible, count


file_data = get_file_data("InputFile")

available_designs = set()
target_towels = set()
for line in file_data:
    if line.__contains__(","):
        for a in line.split(", "):
            available_designs.add(a)
    elif len(line) != 0:
        target_towels.add(line)


part_one = 0
part_two = 0
for towel in target_towels:
    check, number = is_possible(towel)
    if check:
        part_one += 1
    part_two += number

print("Part one answer:", part_one)
print("Part two answer:", part_two)

File: Day19.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Scanner;

public class Day19 {

    public static HashMap<String, TowelResult> cache = new HashMap<String, TowelResult>();

    public static void main(String[] args) {
        ArrayList<String> fileData = getFileData("src/InputFile");
        ArrayList<String> availableTowels = new ArrayList<String>();
        ArrayList<String> targets = new ArrayList<String>();
        for (String line : fileData) {
            if (line.contains(",")) {
                Collections.addAll(availableTowels, line.split(", "));
            } else {
                targets.add(line);
            }
        }
        int partOneAnswer = 0;
        long partTwoAnswer = 0;
        for (String towel : targets) {
            TowelResult t = canMakeTowel(towel, availableTowels);
            boolean check = t.isPossible();
            long combinations = t.getCount();
            if (check) {
                partOneAnswer++;
            }
            partTwoAnswer += combinations;
        }
        System.out.println("Part one answer: " + partOneAnswer);
        System.out.println("Part two answer: " + partTwoAnswer);
    }

    public static TowelResult canMakeTowel(String targetTowel, ArrayList<String> availableTowels) {

        if (cache.keySet().contains(targetTowel)) {
            return cache.get(targetTowel);
        }

        if (targetTowel.length() == 0) {
            TowelResult t = new TowelResult(true, 1);
            cache.put(targetTowel, t);
            return t;
        }

        boolean possible = false;
        long count = 0;
        for (String start : availableTowels) {
            if (targetTowel.indexOf(start) == 0) {
                TowelResult t = canMakeTowel(targetTowel.substring(start.length()), availableTowels);
                boolean check = t.isPossible();
                long combinations = t.getCount();
                if (check) {
                    possible = true;
                    count += combinations;
                }
            }
        }
        TowelResult t = new TowelResult(possible, count);
        cache.put(targetTowel, t);
        return t;
    }

    public static ArrayList<String> getFileData(String fileName) {
        ArrayList<String> fileData = new ArrayList<String>();
        try {
            File f = new File(fileName);
            Scanner s = new Scanner(f);
            while (s.hasNextLine()) {
                String line = s.nextLine();
                if (!line.equals(""))
                    fileData.add(line);
            }
            return fileData;
        }
        catch (FileNotFoundException e) {
            return fileData;
        }
    }
}

class TowelResult {
    private boolean possible;
    private long count;

    public TowelResult(boolean possible, long count) {
        this.possible = possible;
        this.count = count;
    }

    public boolean isPossible() {
        return possible;
    }

    public long getCount() {
        return count;
    }
}
